File: 0053-maximum-subarray/0053-maximum-subarray.rb
# @param {Integer[]} nums
# @return {Integer}
def max_sub_array(nums)
    sum = nums[0]

    n = nums.length
    max_result = sum 

    for i in (1...nums.length) do 
        num = nums[i]
        sum = [sum + num, num].max 
        max_result = [max_result, sum].max 
    end 

    return max_result 
end

File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        n = len(nums)
        result = nums[0]
        curr_sum = 0 

        for i in range(n):
            curr_sum = max(curr_sum + nums[i], nums[i])
            result = max(result, curr_sum)

        return result 

        